File: tutorial/data/project-walkthrough/pkg-r/analysis/code/04_dd_table.R
# 04_dd_table.R
# Professional multi-column DiD table
#
# Presents five regression models side by side in a single LaTeX table:
#   m1 — Baseline TWFE, fatal crashes
#   m2 — TWFE + demographic controls
#   m3 — Southern states only (heterogeneity check)
#   m4 — Non-Southern states only (heterogeneity check)
#   m5 — Alternative outcome: serious crashes
#
# Subgroup models (m3, m4) test whether the policy effect differs by region.
# The alternative outcome (m5) checks whether effects generalise beyond
# fatal crashes to the broader category of serious crashes.
#
# Outputs:
#   analysis/output/tables/dd_table.tex

cat("Running 04_dd_table.R...\n")

# ------------------------------------------------------------------
# Setup — master.R sets these; edit below if running standalone
if (!exists("root")) {
  root     <- "."
  build    <- file.path(root, "build")
  analysis <- file.path(root, "analysis")
  if (!require("pacman")) install.packages("pacman")
  pacman::p_load(data.table, fixest, ggplot2)
}
# ------------------------------------------------------------------

pacman::p_load(data.table, fixest)

# ---- Read data -------------------------------------------------------
dt <- fread(file.path(root, "build/output/analysis_panel.csv"))

# ---- Create variables ------------------------------------------------
if (!"post_treated" %in% names(dt)) {
  dt[, post_treated := fifelse(!is.na(adoption_year) & year >= adoption_year, 1L, 0L)]
}

# South indicator: the Census Bureau defines the South as the 16 states
# (plus DC) in the "South" region. The analysis_panel uses that labeling.
cat("\nRegions in data:\n")
print(dt[, .N, by = region][order(-N)])

dt[, south := as.integer(region == "South")]

cat("\nSample breakdown:\n")
cat("  Total obs.:", nrow(dt), "\n")
cat("  South obs.:", dt[south == 1, .N], "\n")
cat("  Non-South: ", dt[south == 0, .N], "\n")

# ---- Model 1: Baseline TWFE ------------------------------------------
m1 <- feols(
  fatal_crashes ~ post_treated | state_fips + year,
  data = dt,
  vcov = ~state_fips
)

# ---- Model 2: TWFE + controls ----------------------------------------
m2 <- feols(
  fatal_crashes ~ post_treated + log_pop + median_income | state_fips + year,
  data = dt,
  vcov = ~state_fips
)

# ---- Model 3: South only ---------------------------------------------
# Restriction: south == 1
# This tests whether the policy was particularly effective (or ineffective)
# in Southern states, which often have higher baseline crash rates.
m3 <- feols(
  fatal_crashes ~ post_treated + log_pop + median_income | state_fips + year,
  data = dt[south == 1],
  vcov = ~state_fips
)

# ---- Model 4: Non-South only -----------------------------------------
m4 <- feols(
  fatal_crashes ~ post_treated + log_pop + median_income | state_fips + year,
  data = dt[south == 0],
  vcov = ~state_fips
)

# ---- Model 5: Serious crashes outcome --------------------------------
# Uses the same specification as m2 but replaces the outcome.
# If the policy reduces fatalities but not serious crashes (or vice versa),
# that tells us something about the mechanism.
m5 <- feols(
  serious_crashes ~ post_treated + log_pop + median_income | state_fips + year,
  data = dt,
  vcov = ~state_fips
)

# ---- Console output --------------------------------------------------
cat("\n--- Model results ---\n")
for (i in 1:5) {
  cat(sprintf("\nm%d:\n", i))
  print(summary(get(paste0("m", i))))
}

# ---- Export LaTeX table ----------------------------------------------
dir.create(file.path(root, "analysis/output/tables"),
           showWarnings = FALSE, recursive = TRUE)

# Column headers appear above model numbers in a multi-row header.
# We use a list to create a grouped header:
#   Row 1: "Full sample" spanning cols 1-2, "By region" spanning 3-4, "Alt. outcome" col 5
#   Row 2: individual model labels
etable(
  m1, m2, m3, m4, m5,
  title   = "Difference-in-Differences: Robustness Checks",
  headers = list(
    "Full sample" = 2,
    "By region"   = 2,
    "Alt.\\ DV"   = 1
  ),
  dict = c(
    post_treated  = "Post $\\times$ Treated",
    log_pop       = "log(Population)",
    median_income = "Median income",
    fatal_crashes   = "Fatal crashes",
    serious_crashes = "Serious crashes"
  ),
  notes  = "All models include state and year fixed effects. Standard errors clustered by state.",
  file   = file.path(root, "analysis/output/tables/dd_table.tex"),
  replace = TRUE
)

cat("\n  Saved dd_table.tex\n")

File: tutorial/data/project-walkthrough/pkg-python/analysis/code/04_dd_table.py
"""
Analysis Script 04: Multi-Column DD Table

Estimates five DD models side-by-side using pyfixest and compares them
with etable().  This is a common table structure in applied micro papers:
the first columns establish the main result, subsequent columns probe
robustness (controls, subsamples, alternative outcomes).

Models:
  m1 – Baseline: no controls
  m2 – Add time-varying controls (log_pop, median_income)
  m3 – Southern states only
  m4 – Non-Southern states only
  m5 – Alternative outcome (serious_crashes, levels)

All models use state and year fixed effects.  Models m1-m4 cluster SEs
by state; m5 also clusters by state.

Output: analysis/output/tables/dd_table.txt
        analysis/output/tables/dd_table.tex
"""

import io
import os
from contextlib import redirect_stdout
from pathlib import Path

import numpy as np
import pandas as pd

try:
    import pyfixest as pf
except ImportError:
    raise ImportError("pyfixest not installed — run: pip install pyfixest")

ROOT = Path(os.environ.get("PROJECT_ROOT", Path(__file__).parent.parent.parent))

# ─── Paths ────────────────────────────────────────────────────────────────────
input_file  = ROOT / "build/output/analysis_panel.csv"
output_file = ROOT / "analysis/output/tables/dd_table.txt"
output_tex  = ROOT / "analysis/output/tables/dd_table.tex"

print("=" * 60)
print("SCRIPT 04: MULTI-COLUMN DD TABLE")
print("=" * 60)

# ─── Load data ────────────────────────────────────────────────────────────────
print(f"\nReading: {input_file.relative_to(ROOT)}")
df = pd.read_csv(input_file)
print(f"Loaded {len(df):,} state-year observations")

# ─── Construct treatment indicator ────────────────────────────────────────────
# post_treated = 1 for states that adopted the policy AND are past adoption.
df["post_treated"] = (
    (df["policy_adopted"] == 1) & (df["year"] >= df["adoption_year"])
).astype(int)

# ─── Subsets for robustness ───────────────────────────────────────────────────
df_south     = df[df["region"] == "South"].copy()
df_non_south = df[df["region"] != "South"].copy()

print(f"\n  Full sample:      {len(df):,} obs")
print(f"  South only:       {len(df_south):,} obs")
print(f"  Non-South only:   {len(df_non_south):,} obs")

# ─── Model 1: Baseline — no controls ──────────────────────────────────────────
print("\n" + "-" * 60)
print("MODEL 1: Baseline (no controls)")
print("  fatal_crashes ~ post_treated | state_fips + year")
print("-" * 60)

m1 = pf.feols(
    "fatal_crashes ~ post_treated | state_fips + year",
    data=df,
    vcov={"CRV1": "state_fips"},
)
m1.summary()

# ─── Model 2: Add controls ────────────────────────────────────────────────────
print("\n" + "-" * 60)
print("MODEL 2: With controls (log_pop, median_income)")
print("  fatal_crashes ~ post_treated + log_pop + median_income | ...")
print("-" * 60)

m2 = pf.feols(
    "fatal_crashes ~ post_treated + log_pop + median_income | state_fips + year",
    data=df,
    vcov={"CRV1": "state_fips"},
)
m2.summary()

# ─── Model 3: South subsample ─────────────────────────────────────────────────
print("\n" + "-" * 60)
print("MODEL 3: South subsample")
print("  fatal_crashes ~ post_treated | state_fips + year  [South only]")
print("-" * 60)

m3 = pf.feols(
    "fatal_crashes ~ post_treated | state_fips + year",
    data=df_south,
    vcov={"CRV1": "state_fips"},
)
m3.summary()

# ─── Model 4: Non-South subsample ─────────────────────────────────────────────
print("\n" + "-" * 60)
print("MODEL 4: Non-South subsample")
print("  fatal_crashes ~ post_treated | state_fips + year  [Non-South only]")
print("-" * 60)

m4 = pf.feols(
    "fatal_crashes ~ post_treated | state_fips + year",
    data=df_non_south,
    vcov={"CRV1": "state_fips"},
)
m4.summary()

# ─── Model 5: Alternative outcome — serious crashes ───────────────────────────
print("\n" + "-" * 60)
print("MODEL 5: Alternative outcome (serious_crashes)")
print("  serious_crashes ~ post_treated + log_pop | state_fips + year")
print("-" * 60)

m5 = pf.feols(
    "serious_crashes ~ post_treated + log_pop | state_fips + year",
    data=df,
    vcov={"CRV1": "state_fips"},
)
m5.summary()

# ─── Side-by-side comparison ──────────────────────────────────────────────────
print("\n" + "=" * 60)
print("SIDE-BY-SIDE COMPARISON (etable)")
print("=" * 60)
pf.etable([m1, m2, m3, m4, m5])

# ─── Save output ──────────────────────────────────────────────────────────────
def capture(fn):
    buf = io.StringIO()
    with redirect_stdout(buf):
        fn()
    return buf.getvalue()

output_file.parent.mkdir(parents=True, exist_ok=True)

with open(output_file, "w") as f:
    f.write("=" * 80 + "\n")
    f.write("MULTI-COLUMN DIFFERENCE-IN-DIFFERENCES TABLE\n")
    f.write("=" * 80 + "\n\n")

    f.write("MODEL 1: Baseline — no controls\n")
    f.write("  fatal_crashes ~ post_treated | state_fips + year\n")
    f.write("-" * 80 + "\n")
    f.write(capture(m1.summary) + "\n")

    f.write("MODEL 2: With controls (log_pop, median_income)\n")
    f.write("  fatal_crashes ~ post_treated + log_pop + median_income | state_fips + year\n")
    f.write("-" * 80 + "\n")
    f.write(capture(m2.summary) + "\n")

    f.write("MODEL 3: South subsample\n")
    f.write("  fatal_crashes ~ post_treated | state_fips + year  [region == South]\n")
    f.write("-" * 80 + "\n")
    f.write(capture(m3.summary) + "\n")

    f.write("MODEL 4: Non-South subsample\n")
    f.write("  fatal_crashes ~ post_treated | state_fips + year  [region != South]\n")
    f.write("-" * 80 + "\n")
    f.write(capture(m4.summary) + "\n")

    f.write("MODEL 5: Alternative outcome — serious_crashes\n")
    f.write("  serious_crashes ~ post_treated + log_pop | state_fips + year\n")
    f.write("-" * 80 + "\n")
    f.write(capture(m5.summary) + "\n")

    f.write("SIDE-BY-SIDE COMPARISON\n")
    f.write("-" * 80 + "\n")
    f.write(capture(lambda: pf.etable([m1, m2, m3, m4, m5])) + "\n")

# ─── Save LaTeX output ──────────────────────────────────────────────────────
tex = pf.etable([m1, m2, m3, m4, m5], type="tex", labels={
    "post_treated": r"Treatment $\times$ Post",
    "log_pop": "log(Population)",
    "median_income": "Median Income",
    "fatal_crashes": "Fatal Crashes",
    "serious_crashes": "Serious Crashes",
})
with open(output_tex, "w") as f:
    f.write(tex + "\n")

print(f"\nSaved: {output_file.relative_to(ROOT)}")
print(f"Saved: {output_tex.relative_to(ROOT)}")
print("  Models: 5 (baseline, controls, South, Non-South, serious crashes)")
print("  SE type: cluster-robust (CRV1) by state_fips")
